rfcs/038-unify-shared-dependencies.md

```diff
@@ -0,0 +1,112 @@
+- Feature Name: Unify shared dependencies
+- Start Date: 2021-05-22
+- RFC PR: [amundsen-io/rfcs#38](https://github.com/amundsen-io/rfcs/pull/38) (after opening the RFC PR, update this with a link to it and update the file name)
+- Amundsen Issue: [amundsen-io/amundsen#0000](https://github.com/amundsen-io/amundsen/issues/0000) (leave this empty for now)
+
+# Unify shared dependencies
+
+## Summary
+
+Major motivation behind migrating to monorepo was to solve `Dependency Management` issue. However, it still remains unsolved.
+
+This RFC serves as a platform to figure out the best approach for unifying shared dependencies and contains preliminary proposal for how that could be solved.
+
+## Motivation
+
+Following original RFC:
+
+```
+Dependency Management: Python packages are pretty much the same for each proxy, which makes it hard to sync across all the above repositories. As a result most of the time, each proxy is running its own version of the dependency.
+A few examples:
+
+| Package             | amundsenmetadata   | amundsenfrontend  |
+| --------------      |:------------------:| -----------------:|
+| amundsen-common     | >=0.8.1            | ==0.6.0           |
+| flake8              | ==3.5.0            | ==3.8.4           |
+| flake8-tidy-imports | ==1.1.0            | ==4.2.1           |
+
+and many more... This change will put all the packages in just one place, making practically to keep dependencies similar accross verisons, simply because it will be possible to open a single PR that updates multiple packages. Practically speaking this will improve the status quo slightly.
+```
+
+## Guide-level Explanation (aka Product Details)
+
+n/a
+
+## UI/UX-level Explanation
+
+n/a
+
+## Reference-level Explanation (aka Technical Details)
+
+Following steps are proposed to achieve goal of this RFC. 
+
+Multiple times you might find the term `subdirectory`, which refers to following folders:
+- common
+- databuilder
+- frontend
+- metadata
+- search
+
+Steps:
+
+1. Split `requirements.txt` for each subdirectory into `requirements.txt` and `requirements-dev.txt` (there is no need to install flake8 when doing for example `pip install amundsen-metadata` which is currently the case)
+2. Identify libraries that are shared across all (or majority - see `Drawbacks`) of subprojects
+3. Create `requiremnts-common.txt` file in root folder of amundsen repo containing pinned dependencies for shared libraries
+4. Remove libraries listed in `requirements-common.txt` from `requirements.txt` of each subdirectory
+5. Create `requirements-dev.txt` file in root folder of amundsen repo containing pinned dependencies for shared dev libraries (mypy, flake8, etc.)
+6. Remove libraries listed in `requirements-dev.txt` from `requirements.txt` of each subdirectory
+7. Create symbolic link to `requirements-common.txt` and `requirements-dev.txt` in each subdirectory
```

Comment:
Last time I checked, symlinks didn't work on Windows :/ -- do you know if that's been changed?

Reply:
I think git symlinks are supported ![image](https://user-images.githubusercontent.com/5680655/119235059-e6352900-bb30-11eb-8fe5-b3f9b564f78f.png)

Reply:
nice